components/script_injector/renderer/script_injector_render_frame_observer.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/script_injector/renderer/script_injector_render_frame_observer.h"
+
+#include <string>
+#include <utility>
+
+#include "content/public/renderer/render_frame.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_registry.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_script_source.h"
+
+namespace script_injector {
+
+ScriptInjectorRenderFrameObserver::ScriptInjectorRenderFrameObserver(
+    content::RenderFrame* render_frame)
+    : content::RenderFrameObserver(render_frame) {
+  render_frame->GetAssociatedInterfaceRegistry()
+      ->AddInterface<script_injector::mojom::ScriptInjector>(
+          base::BindRepeating(&ScriptInjectorRenderFrameObserver::BindReceiver,
+                              base::Unretained(this)));
```

Comment:
This is a `mojo::PendingAssociatedReceiver`

Reply:
This one is a bit interesting. It's generally only safe to use `base::Unretained` when you own the interface and in this case it's actually owned by the render frame, not `ScriptInjectorRenderFrameObserver` so normally it would not be safe. However, since this is a RenderFrameObserver its lifetime is tied to the render frame. I see both `base::Unretained` and WeakPtr used in this scenario in upstream code, but I think just to be consistent we should use `WeakPtr` here since this class does not own it.